pkg/liqo-controller-manager/resource-request-controller/offerUpdater.go

```diff
@@ -35,118 +31,97 @@ import (
+	ReadResources(clusterID string) corev1.ResourceList
+	// Register sets the component that will be notified of changes.
+	Register(ResourceUpdateNotifier)
+	// RemoveClusterID removes the given clusterID from all internal structures.
+	RemoveClusterID(clusterID string)
+}
 
-// OfferUpdater is a component which wraps all ResourceOffer update logic.
+// OfferUpdater is a component that responds to ResourceRequests with the cluster's resources read from ResourceReader.
 type OfferUpdater struct {
-	queue workqueue.RateLimitingInterface
-	client.Client
-	broadcasterInt            interfaces.ClusterResourceInterface
+	ResourceReader ResourceReader
+	OfferQueue
+
+	client                    client.Client
 	homeCluster               discoveryv1alpha1.ClusterIdentity
 	clusterLabels             map[string]string
 	scheme                    *runtime.Scheme
 	localRealStorageClassName string
 	enableStorage             bool
-}
-
-// Setup initializes all parameters of the OfferUpdater component.
-func (u *OfferUpdater) Setup(cluster discoveryv1alpha1.ClusterIdentity, scheme *runtime.Scheme,
-	broadcaster interfaces.ClusterResourceInterface, k8Client client.Client, clusterLabels map[string]string,
-	localRealStorageClassName string, enableStorage bool) {
-	u.broadcasterInt = broadcaster
-	u.Client = k8Client
-	u.homeCluster = cluster
-	u.scheme = scheme
-	u.clusterLabels = clusterLabels
-	u.queue = workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "Offer update queue")
-	u.localRealStorageClassName = localRealStorageClassName
-	u.enableStorage = enableStorage
-}
-
-// Start runs the OfferUpdate worker.
-func (u *OfferUpdater) Start(ctx context.Context, group *sync.WaitGroup) {
-	defer u.queue.ShutDown()
-	defer group.Done()
-	go u.startRunner(ctx)
-	<-ctx.Done()
-}
+	// currentResources maps the clusters that we intend to offer resources to, to the resource list that we last used
+	// when issuing them a ResourceOffer.
+	currentResources map[string]corev1.ResourceList
+	// updateThresholdPercentage is the change in resources that triggers an update of ResourceOffers.
+	updateThresholdPercentage uint
 
-func (u *OfferUpdater) startRunner(ctx context.Context) {
-	wait.Until(u.run, 2*time.Second, ctx.Done())
+	clusterIdentityCache map[string]discoveryv1alpha1.ClusterIdentity
 }
 
-func (u *OfferUpdater) run() {
-	for u.processNextItem() {
+// NewOfferUpdater constructs a new OfferUpdater.
+func NewOfferUpdater(k8sClient client.Client, homeCluster discoveryv1alpha1.ClusterIdentity,
+	clusterLabels map[string]string, scheme *runtime.Scheme, reader ResourceReader,
+	updateThresholdPercentage uint, localRealStorageClassName string, enableStorage bool) *OfferUpdater {
+	updater := &OfferUpdater{
+		ResourceReader:            reader,
+		client:                    k8sClient,
+		homeCluster:               homeCluster,
+		clusterLabels:             clusterLabels,
+		scheme:                    scheme,
```

Comment:
The scheme can be retrieved from the client, it is not necessary to pass it explicitly